update dependencies and remove node-uuid

SavePointSam opened the pull request:
This update removes the usage of `node-uuid` as it has been deprecated in favor of `uuid`. However, `uuid` no longer has the method `parse` that is used by this library. There has been a re-publish of `parse` using a fork of the original `node-uuid` that strips out everything other than the `parse` and `unparse` methods called `uuid-parse`.

uuid generation was still being used for testing purposes, so `uuid` has been added as a devDependency and related references have been updated to the newer API.

There was also a security vulnerability present in the older version of `istanbul` being used by this project. Updating `istanbul` has patched this without any apparent regressions.

Closes #13.

SavePointSam committed:
fix: replace node-uuid dependency

SavePointSam committed:
fix: update tests and patch security vulnerability

SavePointSam committed:
docs: update examples to recommend uuid over node-uuid

sanlean reviewed (approved):
Excellent solution! Works great for me.

MaerF0x0 commented:
@niahmiah  Any plans for getting this merged? Wanting to move forward with mongoose-uuid

niahmiah commented:
I would merge it, but the tests fail.

SavePointSam committed:
chore: remove deprecation warning from tests

SavePointSam committed:
chore: update coveralls

SavePointSam committed:
chore: update bson

SavePointSam committed:
chore: update mocha and chai

SavePointSam committed:
chore: remove dead code and update tests

SavePointSam committed:
chore: update error message to be more accurate

SavePointSam commented:
After double checking the tests I noticed that there was a few problem dependencies. I went and updated everything. I've also cleaned out a bunch of dead code and updated the tests. The only branch I wasn't sure how to cover came from the use of mongoose internals.

I know this is now a much larger change than was original posted, though it seemed valuable to remove all the duplicate / dead code.

MaerF0x0 commented on index.js:
```diff
@@ -3,127 +3,80 @@
   var b = binary.read(0,len);
-
   var buf = new Buffer(len);
+  var hex = '';
```
For L17-33 ,  Why not use https://www.npmjs.com/package/uuid-parse#uuidparseunparsebuffer-offset    ? 

```javascript
function getter(binary) {
  const string = uuidParse.unparse(bytes); // -> '797ff043-11eb-11e1-80d6-510998755d10'
```

or failing that:  https://nodejs.org/api/buffer.html#buffer_class_method_buffer_from_string_encoding
```javascript
const buf2 = Buffer.from('7468697320697320612074c3a97374', 'hex');
```

SavePointSam replied:
Originally, I was just trying to do the minimum amount of work to get the dependencies updated. I hadn't really scrutinized the implementation. Though, now that we are those are certainly great alternatives. I'll update with a cleaner implementation.

MaerF0x0 commented on test/index.js:
```diff
@@ -17,31 +18,51 @@ ProductSchema.set('toJSON', {getters: true});
+
+  it('should not throw when given anything other than a Buffer or String', function() {
+    var product = new Product({
+      _id: 300,
```
I'm not sure if swallowing validation errors is a good idea?

SavePointSam replied:
The error is never actually thrown. I tried to make it throw many times, but it wouldn't. I believe this has to do with how we are wrapping mongoose's Buffer. I originally wrote the test to make sure it did throw, but the expect is being caught somewhere and a UUID is still output.

What would you suggest we do instead?

MaerF0x0 replied:
I see, ok then maybe a clearer explanation could help, but feel free to leave as is.  👍

stevegoossens commented:
Can this be merged then?
